File: src/dblp_text.py
#!/usr/bin/env python3
# Filter the raw DBLP XML stream and emit one entry per line.
#
# Single pass, regex-based: no XML DOM is built and no external tools are needed.
# Two output formats (--format):
#   text (default) — one grep-friendly line per entry, for the dblp.txt.gz database.
#   sql            — INSERT statements for the SQLite database (see schema.sql).
# DBLP's regular, line-oriented formatting is assumed.  This is the readable Python
# reference; it must stay byte-for-byte in step with dblp_text.go (see `make test`).
import argparse
import os
import re
import sys

# Vendored PyYAML
_VENDORED_YAML = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "vendor", "pyyaml", "lib")
if not os.path.isdir(_VENDORED_YAML):
    sys.exit("dblp_text.py: vendored PyYAML missing -- run `git submodule update --init vendor/pyyaml`")
sys.path.insert(0, _VENDORED_YAML)
import yaml

# ANSI SGR escape codes for --color text output (applied per field in main()).
RESET     = "\x1b[0m"
BOLD      = "\x1b[1m"
UNDERLINE = "\x1b[4m"
GREEN     = "\x1b[32m"
CYAN      = "\x1b[36m"

# The five predefined XML entities plus the named character entities from the DBLP DTD (e.g. &auml; -> ä)
def load_dtd(path):
    ent = {'amp': '&', 'lt': '<', 'gt': '>', 'quot': '"', 'apos': "'"}
    entity_re = re.compile(r'<!ENTITY\s+(\w+)\s+"&#(x?[0-9A-Fa-f]+);"')
    try:
        f = open(path, encoding='utf-8')
    except OSError:
        sys.stderr.write(
            "dblp_text.py: DTD '%s' not found; named entities "
            "(e.g. &auml;) will be left unexpanded\n" % path)
        return ent
    with f:
        for line in f:
            m = entity_re.search(line)
            if not m:
                continue
            name, code = m.group(1), m.group(2)
            cp = int(code[1:], 16) if code.startswith('x') else int(code)
            ent[name] = chr(cp)
    return ent


ENT_RE = re.compile(r'&(#x[0-9A-Fa-f]+|#\d+|[A-Za-z]+);')
def unescape(s, ent):
    if '&' not in s:                         # most fields carry no entity
        return s

    def repl(m):
        e = m.group(1)
        if e.startswith('#x'):
            return chr(int(e[2:], 16))
        if e.startswith('#'):
            return chr(int(e[1:]))
        return ent.get(e, m.group(0))
    return ENT_RE.sub(repl, s)


TAG_RE = re.compile(r'<[^>]+>')
def strip_tags(s):
    return TAG_RE.sub('', s) if '<' in s else s   # most fields carry no inner markup


# Match the title-normalisation used by the SQLite `title_norm` column:
# keep ASCII alphanumerics only, lowercased.
NON_ALNUM_RE = re.compile(r'[^A-Za-z0-9]')
def norm_title(s):
    return NON_ALNUM_RE.sub('', s).lower()


def sql_quote(s):
    # An absent field is None here; coerce to '' so it serialises as an empty SQL string,
    # not the literal "None" that str(None) would produce.
    s = '' if s is None else str(s)
    return "'" + s.replace("'", "''") + "'"


# Return true if the venue is selected: listed explicitly, or "*" (pass everything).
def wanted(vset, venue):
    return '*' in vset or venue in vset


# Start tag of an article/inproceedings with a key; captures (from-tag, kind, venue).
# Venue membership is then a set lookup, not a huge regex alternation.
START_RE = re.compile(r'(<(?:article|inproceedings)\s.*key="(journals|conf)/([^/"]+)/.*)')
# A line up to and including the entry's closing tag.  In DBLP the closing tag of
# one entry and the opening tag of the next often share a line, so closing and
# (re)opening are both checked per line.
CLOSE_RE = re.compile(r'(.*</(?:article|inproceedings)>)')
# Proceedings volume records; captured into the `proceedings` table.  Usually conf/,
# but some workshop volumes are hosted under journals/corr (arXiv/EPTCS), so allow both
# (matching the entry rule and the Go extractor).
PROC_START_RE = re.compile(r'(<proceedings\s.*key="(journals|conf)/([^/"]+)/.*)')
PROC_CLOSE_RE = re.compile(r'(.*</proceedings>)')
YEAR_RE = re.compile(r'<year>(\d+)</year>')
KEY_RE = re.compile(r'key="([^"]*)"')
TYPE_RE = re.compile(r'<(article|inproceedings)\b')
VENUE_RE = re.compile(r'\A(?:journals|conf)/([^/]+)/')
AUTHOR_RE = re.compile(r'<author\b[^>]*>(.*?)</author>', re.S)
EE_RE = re.compile(r'<ee\b[^>]*>(.*?)</ee>', re.S)
FIELD_TAGS = ['journal', 'booktitle', 'series', 'volume', 'number', 'pages', 'title', 'publisher', 'isbn', 'url', 'crossref']
FIELD_RE = {t: re.compile(r'<%s\b[^>]*>(.*?)</%s>' % (t, t), re.S) for t in FIELD_TAGS}

REF_FIELDS = ['journal', 'booktitle', 'series', 'volume', 'number', 'pages']

def text_of(rec, tag, ent):    # first <tag>..</tag>, inner tags stripped, entities unescaped
    m = FIELD_RE[tag].search(rec)
    if not m:
        return None
    return unescape(strip_tags(m.group(1)), ent)


def first_group(rx, s, default=None):  # first capture group of the first match, or `default`
    m = rx.search(s)
    return m.group(1) if m else default


# Preferences: which journals/conferences and which year range survive.
# Returns None when the file is absent or isn't a YAML mapping, so the caller can report it as missing.
def load_config(path):
    try:
        with open(path, encoding='utf-8') as f:
            config = yaml.safe_load(f)
    except OSError:
        return None
    if not isinstance(config, dict):
        return None
    journals = set(config.get('journals') or [])
    confs = set(config.get('conferences') or [])
    journal_names = dict(config.get('journal_names') or {})
    year = config.get('year') or {}
    return journals, confs, journal_names, year.get('lower', 1900), year.get('upper', 2100)


def reference_of(fields):
    return ', '.join(fields[t] for t in REF_FIELDS if fields[t] is not None)


def doi_of(ees):
    for e in ees:
        if 'doi.org' in e:
            return e
    return ees[0] if ees else ''


def main():
    p = argparse.ArgumentParser(add_help=True)
    p.add_argument('--color', action='store_true', help='ANSI coloring (text format)')
    p.add_argument('--format', default='text', help='output format: text (default) or sql')
    p.add_argument('--config', default='config.yaml', help='preference YAML (default: config.yaml)')
    p.add_argument('--dtd', default='dblp.dtd', help='DTD file for entity definitions (default: dblp.dtd)')
    args = p.parse_args()

    cfg = load_config(args.config)
    if cfg is None:
        sys.exit("dblp_text.py: config '%s' not found "
                 "(e.g., try `cp config/sample-se.yaml config/default.yaml`)" % args.config)
    journals, confs, journal_names, lower, upper = cfg
    # A "*" entry passes every venue of that kind (no filtering); see sample-all.yaml.
    if not journals and not confs:
        sys.exit("dblp_text.py: config '%s' selects no venues; list ids under "
                 "journals:/conferences:, or use '- \"*\"' to pass everything" % args.config)

    ent = load_dtd(args.dtd)
    sql_out = args.format == 'sql'

    articles = []
    procs = []          # proceedings volume records (sql format only)
    rec = None          # list of line fragments for the record currently accumulating
    rectype = None      # 'entry' (article/inproceedings) or 'proc' (proceedings)

    for line in sys.stdin:
        if rec is not None:
            close_re = PROC_CLOSE_RE if rectype == 'proc' else CLOSE_RE
            m = close_re.search(line)
            if m:
                record = ''.join(rec) + m.group(1)
                rt = rectype
                rec = None
                rectype = None
                year = first_group(YEAR_RE, record)
                if year is None or lower <= int(year) <= upper:
                    key = first_group(KEY_RE, record, '')
                    if rt == 'proc':
                        procs.append({
                            'key': key,
                            'title': text_of(record, 'title', ent) or '',
                            'booktitle': text_of(record, 'booktitle', ent) or '',
                            'year': int(year or '0'),
                            'publisher': text_of(record, 'publisher', ent) or '',
                            'isbn': text_of(record, 'isbn', ent) or '',
                            'ee': ' '.join(unescape(e, ent) for e in EE_RE.findall(record)),
                            'url': text_of(record, 'url', ent) or '',
                        })
                    else:
                        fields = {t: text_of(record, t, ent) for t in REF_FIELDS}
                        articles.append({
                            'key': key,
                            'type': first_group(TYPE_RE, record, ''),
                            'venue': first_group(VENUE_RE, key, ''),
                            'year': int(year or '0'),
                            'authors': [unescape(strip_tags(a), ent) for a in AUTHOR_RE.findall(record)],
                            'title': re.sub(r'\.$', '', text_of(record, 'title', ent) or ''),
                            'fields': fields,
                            'ee': [unescape(e, ent) for e in EE_RE.findall(record)],
                            'crossref': text_of(record, 'crossref', ent) or '',
                        })
            else:
                rec.append(line)
        # Cheap substring guard before the regex: only record start tags carry a key
        # attribute, so most lines (authors, titles, …) are skipped here.  Venue
        # selection is a set lookup on the captured venue token.  Proceedings are only
        # collected for the sql format (the text database is papers only).
        if rec is None and 'key="' in line:
            m = START_RE.search(line)
            if m and wanted(journals if m.group(2) == 'journals' else confs, m.group(3)):
                rec = [m.group(1)]
                rectype = 'entry'
            elif sql_out:
                m = PROC_START_RE.search(line)
                if m and wanted(journals if m.group(2) == 'journals' else confs, m.group(3)):
                    rec = [m.group(1)]
                    rectype = 'proc'

    out = sys.stdout
    if sql_out:
        out.write('BEGIN;\n')
        # Journal full names from config (abbreviation -> full title); see schema.sql `journals`.
        for abbrev in sorted(journal_names):
            out.write("INSERT INTO journals(abbrev, full_name) VALUES(%s,%s);\n"
                      % (sql_quote(abbrev), sql_quote(journal_names[abbrev])))
        for a in articles:
            f = a['fields']
            cols = [
                sql_quote(a['key']),                  # key
                sql_quote(a['type']),                 # type
                sql_quote(a['venue']),                # venue
                str(a['year']),                       # year
                sql_quote(', '.join(a['authors'])),   # authors
                sql_quote(a['title']),                # title
                sql_quote(norm_title(a['title'])),    # title_norm
                sql_quote(f['journal']),              # journal
                sql_quote(f['booktitle']),            # booktitle
                sql_quote(f['volume']),               # volume
                sql_quote(f['number']),               # number
                sql_quote(f['pages']),                # pages
                sql_quote(doi_of(a['ee'])),           # doi
                sql_quote(' '.join(a['ee'])),         # ee
                sql_quote(a['crossref']),             # crossref
            ]
            out.write("INSERT INTO entries VALUES(%s);\n" % ','.join(cols))
        for pr in procs:
            cols = [
                sql_quote(pr['key']),        # key
                sql_quote(pr['title']),      # title
                sql_quote(pr['booktitle']),  # booktitle
                str(pr['year']),             # year
                sql_quote(pr['publisher']),  # publisher
                sql_quote(pr['isbn']),       # isbn
                sql_quote(pr['ee']),         # ee
                sql_quote(pr['url']),        # url
            ]
            out.write("INSERT INTO proceedings(key,title,booktitle,year,publisher,isbn,ee,url) "
                      "VALUES(%s);\n" % ','.join(cols))
        out.write('COMMIT;\n')
    else:
        articles.sort(key=lambda a: (a['year'], reference_of(a['fields'])))
        for a in articles:
            key = '(%s)' % a['key']
            authors = ', '.join(a['authors'])
            title = a['title']
            ref = reference_of(a['fields'])
            year = a['year']
            doi = doi_of(a['ee'])
            if args.color:
                key = GREEN + key + RESET
                authors = authors + RESET           # authors are unstyled, just terminated
                title = BOLD + title + RESET
                doi = CYAN + UNDERLINE + doi + RESET
            out.write('%s %s: "%s", %s, %s. %s\n' % (key, authors, title, ref, year, doi))


if __name__ == '__main__':
    main()

File: src/dblp_text.go
// dblp_text.go — fast Go counterpart of dblp_text.py (identical I/O: raw DBLP XML
// on stdin, flags --color/--format/--config/--dtd).  --format text (default) emits
// one line per entry; --format sql emits INSERTs for the SQLite database.
// A byte-scanning producer finds entry boundaries (no regexp) and a pool of worker
// goroutines does the per-entry extraction.  Kept in sync with dblp_text.py; run
// `make test` to verify the two agree.
package main

import (
	"bufio"
	"flag"
	"os"
	"regexp"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"sync"

	"go.yaml.in/yaml/v3"
)

const (
	cKey   = "\x1b[32m"
	cTitle = "\x1b[1m"
	cDoi   = "\x1b[36m\x1b[4m"
	cClear = "\x1b[0m"
)

var (
	yearRe   = regexp.MustCompile(`<year>(\d+)</year>`)
	keyRe    = regexp.MustCompile(`key="([^"]*)"`)
	authorRe = regexp.MustCompile(`(?s)<author\b[^>]*>(.*?)</author>`)
	eeRe     = regexp.MustCompile(`(?s)<ee\b[^>]*>(.*?)</ee>`)
	entRe    = regexp.MustCompile(`&(#x[0-9A-Fa-f]+|#\d+|[A-Za-z]+);`)
	tagRe    = regexp.MustCompile(`<[^>]+>`)
)

var refTags = []string{"journal", "booktitle", "series", "volume", "number", "pages"}
var allTags = []string{"journal", "booktitle", "series", "volume", "number", "pages",
	"title", "publisher", "isbn", "url", "crossref"}
var fieldRe = map[string]*regexp.Regexp{}

func init() {
	for _, t := range allTags {
		fieldRe[t] = regexp.MustCompile(`(?s)<` + t + `\b[^>]*>(.*?)</` + t + `>`)
	}
}

// read-only after main() sets them, so safe for concurrent worker use
var (
	entMap         map[string]string
	lowerY, upperY int
	colorOn        bool
	sqlOut         bool
)

func unescape(s string) string {
	if !strings.Contains(s, "&") {
		return s
	}
	return entRe.ReplaceAllStringFunc(s, func(m string) string {
		e := m[1 : len(m)-1]
		switch {
		case strings.HasPrefix(e, "#x"):
			cp, _ := strconv.ParseInt(e[2:], 16, 32)
			return string(rune(cp))
		case e[0] == '#':
			cp, _ := strconv.ParseInt(e[1:], 10, 32)
			return string(rune(cp))
		default:
			if v, ok := entMap[e]; ok {
				return v
			}
			return m
		}
	})
}

func stripTags(s string) string {
	if !strings.Contains(s, "<") {
		return s
	}
	return tagRe.ReplaceAllString(s, "")
}

func textOf(rec string, re *regexp.Regexp) (string, bool) {
	m := re.FindStringSubmatch(rec)
	if m == nil {
		return "", false
	}
	return unescape(stripTags(m[1])), true
}

func sqlQuote(s string) string { return "'" + strings.ReplaceAll(s, "'", "''") + "'" }

// normTitle mirrors dblp_text.py's title_norm: keep ASCII alphanumerics only,
// lowercased.  ASCII-only on purpose so every extractor and the query side normalise identically.
func normTitle(s string) string {
	var b strings.Builder
	for i := 0; i < len(s); i++ {
		switch c := s[i]; {
		case c >= 'A' && c <= 'Z':
			b.WriteByte(c + 32)
		case (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9'):
			b.WriteByte(c)
		}
	}
	return b.String()
}

func typeOf(record string) string {
	if strings.HasPrefix(record, "<inproceedings") {
		return "inproceedings"
	}
	if strings.HasPrefix(record, "<proceedings") {
		return "proceedings"
	}
	return "article"
}

func venueOf(key string) string {
	rest := ""
	if r, ok := strings.CutPrefix(key, "journals/"); ok {
		rest = r
	} else if r, ok := strings.CutPrefix(key, "conf/"); ok {
		rest = r
	} else {
		return ""
	}
	if i := strings.IndexByte(rest, '/'); i >= 0 {
		return rest[:i]
	}
	return rest
}

type entry struct {
	year int
	ref  string
	out  string
}

// closeCapture returns the line up to and including the rightmost record close tag
// (</article>, </inproceedings> or </proceedings>).  A record contains only its own
// close tag, so searching all three is safe; "</proceedings>" is not a substring of
// "</inproceedings>".
func closeCapture(line string) (string, bool) {
	end := -1
	for _, tag := range []string{"</article>", "</inproceedings>", "</proceedings>"} {
		if i := strings.LastIndex(line, tag); i >= 0 {
			if e := i + len(tag); e > end {
				end = e
			}
		}
	}
	if end < 0 {
		return "", false
	}
	return line[:end], true
}

// startMatch finds the leftmost <article / <inproceedings start tag whose key is
// in a wanted venue, and returns the slice from that tag to end of line (mirrors
// `(<(?:article|inproceedings)\s.*key="(journals|conf)/([^/"]+)/.*)` + set lookup).
func startMatch(line string, journals, confs map[string]bool) (string, bool) {
	start := -1
	if i := strings.Index(line, "<article "); i >= 0 {
		start = i
	}
	if i := strings.Index(line, "<inproceedings "); i >= 0 && (start < 0 || i < start) {
		start = i
	}
	if sqlOut { // proceedings volumes go to the sql output only
		if i := strings.Index(line, "<proceedings "); i >= 0 && (start < 0 || i < start) {
			start = i
		}
	}
	if start < 0 {
		return "", false
	}
	ki := strings.Index(line[start:], `key="`)
	if ki < 0 {
		return "", false
	}
	v := line[start+ki+len(`key="`):]
	ok := false
	if rest, found := strings.CutPrefix(v, "journals/"); found {
		if e := strings.IndexAny(rest, `/"`); e >= 0 && rest[e] == '/' {
			ok = journals["*"] || journals[rest[:e]] // "*" passes every journal
		}
	} else if rest, found := strings.CutPrefix(v, "conf/"); found {
		if e := strings.IndexAny(rest, `/"`); e >= 0 && rest[e] == '/' {
			ok = confs["*"] || confs[rest[:e]] // "*" passes every conference
		}
	}
	if !ok {
		return "", false
	}
	return line[start:], true
}

func extract(record string) (entry, bool) {
	var year int
	hasYear := false
	if m := yearRe.FindStringSubmatch(record); m != nil {
		hasYear = true
		year, _ = strconv.Atoi(m[1])
	}
	if hasYear && (year < lowerY || year > upperY) {
		return entry{}, false
	}
	key := ""
	if m := keyRe.FindStringSubmatch(record); m != nil {
		key = m[1]
	}
	typ := typeOf(record)
	field := func(t string) string { v, _ := textOf(record, fieldRe[t]); return v }
	var ees []string
	for _, m := range eeRe.FindAllStringSubmatch(record, -1) {
		ees = append(ees, unescape(m[1]))
	}

	if typ == "proceedings" {
		out := "INSERT INTO proceedings(key,title,booktitle,year,publisher,isbn,ee,url) VALUES(" +
			sqlQuote(key) + "," + // key
			sqlQuote(field("title")) + "," + // title
			sqlQuote(field("booktitle")) + "," + // booktitle
			strconv.Itoa(year) + "," + // year
			sqlQuote(field("publisher")) + "," + // publisher
			sqlQuote(field("isbn")) + "," + // isbn
			sqlQuote(strings.Join(ees, " ")) + "," + // ee
			sqlQuote(field("url")) + ");" // url
		return entry{year, "", out}, true
	}

	var authors []string
	for _, m := range authorRe.FindAllStringSubmatch(record, -1) {
		authors = append(authors, unescape(stripTags(m[1])))
	}
	title := ""
	if v, ok := textOf(record, fieldRe["title"]); ok {
		title = strings.TrimSuffix(v, ".")
	}
	doi := ""
	for _, e := range ees {
		if strings.Contains(e, "doi.org") {
			doi = e
			break
		}
	}
	if doi == "" && len(ees) > 0 {
		doi = ees[0]
	}
	auth := strings.Join(authors, ", ")

	if sqlOut {
		out := "INSERT INTO entries VALUES(" +
			sqlQuote(key) + "," + // key
			sqlQuote(typ) + "," + // type
			sqlQuote(venueOf(key)) + "," + // venue
			strconv.Itoa(year) + "," + // year
			sqlQuote(auth) + "," + // authors
			sqlQuote(title) + "," + // title
			sqlQuote(normTitle(title)) + "," + // title_norm
			sqlQuote(field("journal")) + "," + // journal
			sqlQuote(field("booktitle")) + "," + // booktitle
			sqlQuote(field("volume")) + "," + // volume
			sqlQuote(field("number")) + "," + // number
			sqlQuote(field("pages")) + "," + // pages
			sqlQuote(doi) + "," + // doi
			sqlQuote(strings.Join(ees, " ")) + "," + // ee
			sqlQuote(field("crossref")) + ");" // crossref
		return entry{year, "", out}, true
	}

	var refs []string
	for _, t := range refTags {
		if v, ok := textOf(record, fieldRe[t]); ok {
			refs = append(refs, v)
		}
	}
	ref := strings.Join(refs, ", ")
	yr := strconv.Itoa(year)
	var out string
	if colorOn {
		out = cKey + "(" + key + ")" + cClear + " " + auth + cClear + ": \"" +
			cTitle + title + cClear + "\", " + ref + ", " + yr + ". " + cDoi + doi + cClear
	} else {
		out = "(" + key + ") " + auth + ": \"" + title + "\", " + ref + ", " + yr + ". " + doi
	}
	return entry{year, ref, out}, true
}

// loadConfig reads the preference YAML with go.yaml.in/yaml/v3: which journals/conferences
// and which year range survive.  Returns an error when the file can't be read or isn't a
// YAML mapping, so the caller can report it as missing.  Mirrors dblp_text.py's load_config.
func loadConfig(path string) (journals, confs map[string]bool, journalNames map[string]string, lower, upper int, err error) {
	journals, confs = map[string]bool{}, map[string]bool{}
	journalNames = map[string]string{}
	lower, upper = 1900, 2100
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}
	var cfg struct {
		Journals     []string                  `yaml:"journals"`
		Conferences  []string                  `yaml:"conferences"`
		JournalNames map[string]string         `yaml:"journal_names"`
		Year         struct{ Lower, Upper int } `yaml:"year"`
	}
	cfg.Year.Lower, cfg.Year.Upper = lower, upper // defaults kept when the year keys are absent
	if err = yaml.Unmarshal(data, &cfg); err != nil {
		return
	}
	for _, j := range cfg.Journals {
		journals[j] = true
	}
	for _, c := range cfg.Conferences {
		confs[c] = true
	}
	for k, v := range cfg.JournalNames {
		journalNames[k] = v
	}
	lower, upper = cfg.Year.Lower, cfg.Year.Upper
	return
}

func loadDTD(path string) map[string]string {
	ent := map[string]string{"amp": "&", "lt": "<", "gt": ">", "quot": "\"", "apos": "'"}
	f, err := os.Open(path)
	if err != nil {
		os.Stderr.WriteString("dblp(go): DTD '" + path + "' not found; named entities left unexpanded\n")
		return ent
	}
	defer f.Close()
	re := regexp.MustCompile(`<!ENTITY\s+(\w+)\s+"&#(x?[0-9A-Fa-f]+);"`)
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		m := re.FindStringSubmatch(sc.Text())
		if m == nil {
			continue
		}
		code := m[2]
		var cp int64
		if code[0] == 'x' {
			cp, _ = strconv.ParseInt(code[1:], 16, 32)
		} else {
			cp, _ = strconv.ParseInt(code, 10, 32)
		}
		ent[m[1]] = string(rune(cp))
	}
	return ent
}

func main() {
	color := flag.Bool("color", false, "ANSI coloring (text format)")
	format := flag.String("format", "text", "output format: text or sql")
	config := flag.String("config", "config.yaml", "preference YAML")
	dtd := flag.String("dtd", "dblp.dtd", "DTD for entity definitions")
	flag.Parse()

	journals, confs, journalNames, lower, upper, err := loadConfig(*config)
	if err != nil {
		os.Stderr.WriteString("dblp(go): config '" + *config + "' not found (e.g., try `cp config/sample-se.yaml config/default.yaml`)\n")
		os.Exit(1)
	}
	if len(journals) == 0 && len(confs) == 0 {
		os.Stderr.WriteString("dblp(go): config '" + *config + "' selects no venues; list ids under journals:/conferences:, or use '- \"*\"' to pass everything\n")
		os.Exit(1)
	}
	entMap = loadDTD(*dtd)
	lowerY, upperY = lower, upper
	colorOn = *color
	sqlOut = *format == "sql"

	const batchSize = 256
	records := make(chan []string, 2*runtime.NumCPU())
	nw := runtime.NumCPU()
	results := make([][]entry, nw)
	var wg sync.WaitGroup
	for w := 0; w < nw; w++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			var local []entry
			for batch := range records {
				for _, rec := range batch {
					if e, ok := extract(rec); ok {
						local = append(local, e)
					}
				}
			}
			results[id] = local
		}(w)
	}

	// producer: scan boundaries, emit raw record strings in batches
	sc := bufio.NewScanner(os.Stdin)
	sc.Buffer(make([]byte, 1<<20), 64<<20)
	var recLines []string
	active := false
	batch := make([]string, 0, batchSize)
	for sc.Scan() {
		line := sc.Text()
		if active {
			if capt, ok := closeCapture(line); ok {
				recLines = append(recLines, capt)
				batch = append(batch, strings.Join(recLines, ""))
				recLines = recLines[:0]
				active = false
				if len(batch) >= batchSize {
					records <- batch
					batch = make([]string, 0, batchSize)
				}
			} else {
				recLines = append(recLines, line, "\n")
			}
		}
		if !active && strings.Contains(line, `key="`) {
			if capt, ok := startMatch(line, journals, confs); ok {
				recLines = append(recLines, capt)
				active = true
			}
		}
	}
	if len(batch) > 0 {
		records <- batch
	}
	close(records)
	wg.Wait()

	var entries []entry
	for _, r := range results {
		entries = append(entries, r...)
	}
	w := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer w.Flush()
	if sqlOut {
		w.WriteString("BEGIN;\n")
		// Journal full names from config (abbreviation -> full title); see schema.sql `journals`.
		names := make([]string, 0, len(journalNames))
		for k := range journalNames {
			names = append(names, k)
		}
		sort.Strings(names)
		for _, k := range names {
			w.WriteString("INSERT INTO journals(abbrev, full_name) VALUES(" +
				sqlQuote(k) + "," + sqlQuote(journalNames[k]) + ");\n")
		}
		for _, e := range entries {
			w.WriteString(e.out)
			w.WriteByte('\n')
		}
		w.WriteString("COMMIT;\n")
		return
	}
	sort.SliceStable(entries, func(i, j int) bool {
		if entries[i].year != entries[j].year {
			return entries[i].year < entries[j].year
		}
		return entries[i].ref < entries[j].ref
	})
	for _, e := range entries {
		w.WriteString(e.out)
		w.WriteByte('\n')
	}
}
